stan/math/fwd/fun/tcrossprod.hpp

```diff
@@ -1,20 +1,22 @@
 #ifndef STAN_MATH_FWD_FUN_TCROSSPROD_HPP
 #define STAN_MATH_FWD_FUN_TCROSSPROD_HPP
 
+#include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/fwd/fun/multiply.hpp>
 
 namespace stan {
 namespace math {
 
-template <typename T, int R, int C>
-inline Eigen::Matrix<fvar<T>, R, R> tcrossprod(
-    const Eigen::Matrix<fvar<T>, R, C>& m) {
+template <typename T, require_eigen_vt<is_fvar, T>* = nullptr>
+inline Eigen::Matrix<value_type_t<T>, T::RowsAtCompileTime,
+                     T::RowsAtCompileTime>
```

Comment:
fyi everywhere T is an Eigen Matrix would be good to have a nicer name like `EigMat`

Reply:
Done. I know about template names, but they are not very high on my priority list, so I tend to forget to change them. You can either keep reminding me about every single instance I miss or we can leave that for a separate series of PRs.

Reply:
One of the return statements is invoking default constructor, so compiler can not deduce `auto`.

Reply:
Except at moment the function will take any eigen type as an input but can only return ```Eigen::Matrix``` types (which could cause some compile errors down the line). You could restrict the inputs to only take matrix types, or since the result will need to be evaluated anyway, this could be written to be more flexible:
```
template <typename T, require_eigen_vt<is_fvar, T>* = nullptr>
inline auto tcrossprod(const T& m) {
  using T_return = decltype(multiply(m, m.transpose()));
  if (m.rows() == 0) {
    return plain_type_t<T_return>{};
  }
  return multiply(m, m.transpose()).eval();
}
```

Reply:
Many functions still assume input is derived from `Eigen::MatrixBase` and will not work with `Eigen::ArrayBase`-derived types. Or are you thinking about some other type this would fail with?

Reply:
No it was just ```ArrayBase``` that I was thinking of, so if that's a non-issue then feel free to ignore